### ISSUE TITLE
Would it be possible to make this code work natively for CUDA pytorch tensors?

### ISSUE THREAD

**Issue by mm04926412** [fwilliams/point-cloud-utils#14]
I appreciate it might be difficult to program but I'm currently creating a loss function for my autoencoder by turning a pytorch tensor into numpy, calculating indicies of nearest neighbour pairs and then just indexing the tensor to get a differentiable loss function from it.

Would it be possible to make this algorithm natively compatible with pytorch cuda tensors to immediatly get a differentiable loss function between two point clouds?

**Comment by fwilliams** [comment]
Hi there,

I appreciate the desire to have this functionality natively integrated with PyTorch (I'm mostly using this library in machine learning applications). 

Unfortunately, it would be a very large amount of effort to deeply integrate this library with PyTorch and I don't have time to do it. To make things work natively with CUDA tensors would essentially require a full rewrite of this library.

Other than nearest neighbor, what other functionality do you need?

**Comment by mm04926412** [comment]
Pretty much just creating loss functions with the distance metrics so I have permutation invariant back propogation. I'm currently trying to get some code working that pulls the numpy array out and passes indices to a pytorch sum layer - I'm pretty new to using pytorch but is there a convinient way to make a "convert to numpy as the pass the indicies to pytorch" paradigm beyond what will probably be my own naive approach

**Comment by fwilliams** [comment]
Not sure if it's what you're looking for but for a differentiable permutation invariant loss between point clouds you can try [Sinkhorn Distances](https://arxiv.org/abs/1306.0895) (it's a regularized optimal transport loss). 
I have a PyTorch implementation of the sinkhorn loss with code examples in [my other library](https://github.com/fwilliams/fml). 

For your nearest neighbor based loss, you can index a PyTorch tensor with a numpy array of integers (you may have to convert it to a torch long tensor `torch.from_numpy(indices).to(torch.long)`). You won't be able to backprop through the index array of course (they're not computed with PyTorch and they're in a discontinuous space of integers). 

Something to note though is that the set of k nearest neighbors will change discontinuously as the points move. This is okay so long as the function you're indexing changes continuously (which is the case if each point contributes a continuous function of distance to the loss so that two points contribute the same value if they're at the same distance).

Best of luck with everything and let me know if you have more questions!

**Comment by mm04926412** [comment]
Thank you this is incredibly helpful, the final thing I'm aiming for is a point cloud autoencoder where the loss function doesn't care about the order, rotation or the translation vs the input so that only the... I'm not sure of the precise language, the 'concept' of the structure is in the latent space. The point clouds I'm using are crystal structures of atoms where the point cloud represents a ground truth rather than a sampling of a shape or something.

Is the sinkhorn distance rotation invariatn? Is it possible to make it such?

**Comment by fwilliams** [comment]
If you have two point clouds `A` and `B`, and you apply the same rotation to both of them, then the sinkhorn distance won't change. So in that sense it is rotation invariant. 

Something to note about sinkhorn distances is that they're relatively expensive to backprop through (you're backpropping through a big unrolled for loop). If you run into performance issues you can take the ouput correspondence matrix of sinkhorn and "project" it to a permutation (set the max in each row to 1 and everything else to zero and then compute a permuted mean square loss). This is almost equivalent and has worked well for me in the past.
